src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Logging/OutputWindowLogger.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Logging;
+
+[Shared]
+[Export(typeof(OutputWindowLogger))]
+internal class OutputWindowLogger : ILogger
+{
+    private const LogLevel MinimumLogLevel = LogLevel.Warning;
+    private readonly OutputPane _outputPane;
+
+    [ImportingConstructor]
+    public OutputWindowLogger(JoinableTaskContext joinableTaskContext)
+    {
+        _outputPane = new OutputPane(joinableTaskContext);
+    }
+
+    [Obsolete("Exists only for Mock.")]
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+    internal OutputWindowLogger()
+#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+    {
+    }
+
+    public IDisposable BeginScope<TState>(TState state)
+    {
+        throw new NotImplementedException();
+    }
+
+    public bool IsEnabled(LogLevel logLevel)
+    {
+        return logLevel >= MinimumLogLevel;
+    }
+
+    public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
+    {
+        if (IsEnabled(logLevel))
+        {
+            _outputPane.WriteLine(formatter(state, exception));
+            if (exception is not null)
+            {
+                _outputPane.WriteLine(exception.ToString());
+            }
+        }
+    }
+
+    private class OutputPane
+    {
+        private static readonly Guid s_outputPaneGuid = new("BBAFF416-4AF5-41F2-9F93-91F283E43C3B");
+
+        private readonly JoinableTaskContext _threadingContext;
+        private readonly IServiceProvider _serviceProvider;
+        private IVsOutputWindowPane? _doNotAccessDirectlyOutputPane;
+
+        public OutputPane(JoinableTaskContext threadingContext)
+        {
+            _threadingContext = threadingContext;
+            _serviceProvider = ServiceProvider.GlobalProvider;
+        }
+
+        public void WriteLine(string value)
+        {
+            WriteLineInternal(value);
+        }
+
+        private void WriteLineInternal(string value)
+        {
+            var pane = GetPane();
+            if (pane is null)
+            {
+                return;
+            }
+
+            pane.OutputStringThreadSafe(value + Environment.NewLine);
```

Comment:
The [docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsoutputwindowpane.outputstringthreadsafe?view=visualstudiosdk-2022) say to try the non-message-pump version of this first, then fallback to this if that fails.

Here is how I did it in Roslyn, after getting a Prism bug about thread crimes:
https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/Core/Def/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs#L66-L73

Reply:
Oh, we should tell Roslyn then cause I [took this from their copy](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/VisualStudioDiagnosticsToolWindow/Loggers/OutputWindowLogger.cs#L82) whole-cloth lol

I'll file an issue against Roslyn and update this to match that pattern.

Reply:
https://github.com/dotnet/roslyn/issues/66397

Reply:
err.. did you create that in the wrong repo? The roslyn one does the _right_ thing, this one should copy it 😛

Reply:
@davidwengier is correct. We wrote similar code in the WinForms Designer:

```C#
// Visual Studio prefers writing to an IVsOutputWindowPane using the "no pump" semantics
// to print a message without any message pumping.
//
// For more detail, see:
// https://docs.microsoft.com/dotnet/api/microsoft.visualstudio.shell.interop.ivsoutputwindowpanenopump#remarks

if (outputPane is IVsOutputWindowPaneNoPump outputPaneNoPump)
{
    outputPaneNoPump.OutputStringNoPump(text);
}
else if (outputPane is not null)
{
    HResult hresult = outputPane.OutputStringThreadSafe(text);
    hresult.ThrowIfFailure();
}
```

Reply:
Discussed offline with @davidwengier. To clarify, Roslyn uses this API in multiple places. The [one that David linked/wrote](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/Core/Def/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs#L66-L73) is correct according to the docs. But I happened to have [used a different class](https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/VisualStudioDiagnosticsToolWindow/Loggers/OutputWindowLogger.cs#L82) when creating this.